### DISCUSSION TITLE
What is the best way to debug pyg?

### DISCUSSION THREAD

**Discussion by RessCris2** [pyg-team/pytorch_geometric#4276] [Q&A]
Hello, I don't really know what NeighborLoader is doing, especially the way hetero_neighbor_sample does, I'm not sure what is the stragtegy it used to sample form bi-graph. I checked the code, it's written from C++ in torch_sparse , I'm wondering how can i debug it, to see step by step how it works?

I'm trying to install pyg in debugging way, I dont know if it works...
Any suggestions?Thanks!

**Comment by rusty1s** [comment]
It is super tricky to debug since we cannot use any Python debugger to debug the C++ code. To debug the C++ code in `torch-sparse`, you would have to build the library from source and setup a C++ testing suite (our current test suite is in Python).

The easiest way would be to create a simple graph example in Python, and modify the source code of `torch-sparse` in order to some traditional `print` debugging.
